### PR TITLE
Major Refactor and Clean up

### LINKED ISSUE #2: Support inverse case
This plugin currently raises an alert when the specified regexp is matched.

We understand from some customers that they have a use for the inverse case, meaning they want to be alerted when specified patterns are not present in the log.

### LINKED ISSUE #10: Add support for regex paths
It would be helpful to be able to specify a regex path/multiple files to look at, specifically for applications that write logs to dated folders.  For example:

`C:\Program Files\Application\logs\YYYY-MM-DD\log.txt`

Being able to present the check configuration as the following would be nice.

`sensu-check-log -log "C:\Program Files\Application\logs\*\*.txt" -match ERROR -state "C:\etc\sensu\log_state\Application"`

This was previously accomplished with a modified version of the log plugin for Sensu Core, but given the current Ruby runtime does not support Windows that guy no longer functions.

### LINKED ISSUE #11: Provide error feedback if file is missing.

### PR DESCRIPTION
**Pull request by jspaleta**
# Refactored Elements
1. Using sensu plugin SDK
2. Extend State object to keep up with match condition used modification time and last process time.
3. Extend result object to track file that matches.

# New Features
## Inverse matching option
Unfortunately the regexp package doesn't support the `!` RE flag syntax, but inverse match can be implemented with a global flag to switch whether no matches causes an alert event to be generated.

## Supports Log file regexp to build list of files in addition to single file support
file regexp is compiled into a list of files.
all files are processed into a single event, with output indicating which files had a match

##  Throws file access errors as non-zero status for calling check
Example: missing log file will result in non-zero status in calling check not in the generated event.
 
## New support for disabling event generation
If event generation disabled, event on stdin is not required.

## Generated event check name template
check-name no longer hard coded, now capable of using either check token substitution or golang templating as supported by the SDK.

## Updated log file state handling
1. Use of state directory and one state file per log file (this can be changed to db in later release)
1. Add automatic offset state reset logic to handle log rotation situations where log file is truncated to 0.
1. Add state protection if match argument changes, to guard against concurrent runs of same command with different match conditions using the same state files.

# Closes
closes #2 
closes #10 
closes #11

### CODE REVIEW

File: main.go
Comment (echlebek) [inline main.go:80]: Required if --log-file-expr _not_ used, right?

File: main.go
Comment (echlebek) [inline main.go:438]: On systems that have atime enabled, the modification time will be whenever the log file was opened, so I think this will be wrong in those cases.

Luckily, you don't actually need to know the modification time, you only need to know the creation time, and that should reset when the log is rotated. I think you can probably remove LastTime and ModTime and add CreatedTime to the State type without losing any functionality.

File: main.go
Comment (echlebek) [inline main.go:205]: It's a bit risky to use large integers in JSON because other tools will interpret them as floats, like jq. It would be better to store RFC-3339/ISO-8601 formatted datetimes anyways, because then we can capture things like TZ.

File: main.go
Comment (echlebek) [inline main.go:207]: It doesn't make sense to me that these fields are represented here, since they are supplied by the tool. Can you explain?

**Review by echlebek — CHANGES_REQUESTED** [review]
I'm getting some good and bad vibes from the PR. On the one hand, the change adds significant complexity to the tool. The number of flags has doubled, and a lot of logic has been added. The CLI of the tool is now completely different, and anyone using the old tool will have to update their use of it.

On the other hand you've done a pretty darn good job and I like the documentation improvements. The change seems mostly correct, with some issues that I've spotted, and likely some more than I haven't. Those holes can get filled over time.

Test coverage is poor, with main.go only enjoying 21% coverage. Tests should be added to cover the new functionality. Given that the complexity has increased, I think high test coverage will be necessary to prove that everything works, and to keep regressions from happening.

Some of the functions could also benefit from refactoring, and I think writing unit tests for them will help with that.

There's two main issues that the tool has right now that I think should be resolved before we merge the PR. The first issue is that all the log files stay open for the duration of the tool's run. That could get messy on big lists of files and it's unnecessary, so let's fix it. The second issue is the modtime issue that I mentioned below.

I'm confident that you'll take responsibility for maintaining the tool in the direction it's heading with these changes, so I'll be happy to approve the PR once the testing firms up, and we sort out the two main issues with the tool.

The next obvious maintenance step for the tool is to allow logs to be processed in parallel, and with that in mind I'm not sure the flag that lets users set the max number of analyzers is still good to have. Let's think about that one before committing to it.

File: main.go
Comment (jspaleta) [reply main.go:376]: The problem is, this wipes out the match state in the generated events. Because the generated events are by design not resolvable... if you do a missing file critical generated event.. you won't ever automatically resolve that if the file comes back with no matches.  
The place for this sort of error is in the calling event because it doesn't break the auto-resolve pattern.

File: main.go
Comment (jspaleta) [reply main.go:516]: uhm... pretty sure if i dont expliclty have the new line... the outputEvent contents get smushed on the same line.  It's just stylist for diagnostic output not a big deal either way.

File: main.go
Comment (jspaleta) [reply main.go:80]: No required if --log-file-expr used.. set the file system base directory from which the rexpr will walk the directory tree.

log-path  "/var/log"  for example  would restrict your regexp to just walk the /var/log/ directory tree

File: main.go
Comment (jspaleta) [reply main.go:438]: Let's be clear when you say creation time.. do you mean atime,mtime or ctime and ctime means change time.  
I don't think unix technically defines a create time.  Some filesystems may track creation time, but I don't think its universal.

In fact it looks like golang just exposes a ModTime() in its os agnostic package for dealing with file information. 
https://pkg.go.dev/io/fs#FileInfo

File: main.go
Comment (jspaleta) [reply main.go:205]: I consider the state file to be an internal implementation detail. it just need to be monotonic like a unix epoch and does not need to be timezone aware or even human readable. It just need to be monotonically increasing.  
In fact I might as well just change the state file to a binary file and not use json at all, probably faster just to read a binary file.

I dislike using the iso standard for strictly computational time keeping because human parsable timestamps continually require external data updating for correct interpretation for corner-cases like positive and negative leap seconds.  There's no need for that.

File: main.go
Comment (jspaleta) [reply main.go:207]: Actually the regexp golang implementation doesn't allow for inverse match syntax. The regexp package implements a subset of the full re syntax.. and inversion isn't part of the supported subset. Ugh.. i was that was not the case, but from my testing, I couldn't make inverse matching work.. regexp threw errors... even in golang playground testing.

File: event.go
Comment (jspaleta) [reply event.go]: resolved via additional commit

File: main.go
Comment (jspaleta) [reply main.go]: resolved in additional commit

File: main.go
Comment (echlebek) [reply main.go:205]: Binary file with seconds since epoch sounds reasonable. Perhaps use `encoding/gob` for convenience.

File: main.go
Comment (jspaleta) [reply main.go]: This feels like something the linter should be linting.

resolved in new commit.

File: main.go
Comment (jspaleta) [reply main.go:80]: would setting a default like /var/log  make more sense?

File: main.go
Comment (jspaleta) [reply main.go:207]: Oh in the state file?  Oh that's because if you change the matching condition you want you've basically invalidated the stored state.
Let's say you create a check and want to match for `Error` and your regexp is case sensitive when you really wanted to match for `error` case-insensitive.
You go back and edit the MatchExpr accordingly.  
You just invalided the state file because you need to reprocess the log file again from the beginning for the new match.

File: main.go
Comment (jspaleta) [reply main.go]: I've set the default for --log-path to /var/log  and updated textual description for the option to be a little clearer.

File: main.go
Comment (jspaleta) [reply main.go]: resolved in additional commit.

File: main.go
Comment (jspaleta) [reply main.go]: refactored in new commit.  The processLogFile function created to handle the defered closure in the for loop

File: main.go
Comment (jspaleta) [reply main.go]: After adding more testing, removed this time fields from the state.
There's a corner case with the state file that we can't seem to handle. 

If the file is rotated and filled again quickly we might not be able to detect that its a new file and the offset needs to be set back to 0.

File: main.go
Comment (jspaleta) [reply main.go]: removed all logic involving state.LastTime and stateModTIme.

**Comment by jspaleta** [comment]
@echlebek 
I've significantly upped the test coverage in main.go  except for some hard to produce error conditions.

**Comment by echlebek** [comment]
@jspaleta ready for a new round of review?

**Comment by jspaleta** [comment]
@echlebek 
yes.

File: main.go
Comment (jspaleta) [reply main.go]: removed from state file in latest commits. 
Logic now just checks to see if file got smaller since last run. If so it resets file read offset to 0.

File: main.go
Comment (echlebek) [reply main.go:528]: Nope https://play.golang.org/p/1kTcEuMvv1U

**Review by echlebek — APPROVED** [review]
```
                              __    __    __
                             |==|  |==|  |==|
                           __|__|__|__|__|__|_
                        __|___________________|___
                     __|__[]__[]__[]__[]__[]__[]__|___
                    |............................o.../
                    \.............................../
               hjw_,~')_,~')_,~')_,~')_,~')_,~')_,~')/,~')_
```

**Pull request merged**
